components/nifosg/nifloader.cpp

```diff
@@ -1030,6 +1030,78 @@ namespace NifOsg
             }
         }
 
+        class PreciseLeafDepthCullCallback : public osg::DrawableCullCallback{
+            inline osgUtil::CullVisitor::value_type distance(const osg::Vec3& coord,const osg::Matrix& matrix) const
+            {
+                return -((osgUtil::CullVisitor::value_type)coord[0]*(osgUtil::CullVisitor::value_type)matrix(0,2)+(osgUtil::CullVisitor::value_type)coord[1]*(osgUtil::CullVisitor::value_type)matrix(1,2)+(osgUtil::CullVisitor::value_type)coord[2]*(osgUtil::CullVisitor::value_type)matrix(2,2)+matrix(3,2));
+            }
+
+            virtual bool cull(osg::NodeVisitor* nv, osg::Drawable* drawable, osg::RenderInfo* renderInfo) const{
+                osgUtil::CullVisitor * cv = nv -> asCullVisitor();
+                osg::RefMatrix& matrix = *cv->getModelViewMatrix();
+                const osg::BoundingBox& bb = drawable->getBoundingBox();
+                if (drawable->isCullingActive() && cv->isCulled(bb)) return true;
+
+
+                if (cv->getComputeNearFarMode() && bb.valid())
+                {
+                    if (!cv->updateCalculatedNearFar(matrix,*drawable,false)) return true;
+                }
+
+                // need to track how push/pops there are, so we can unravel the stack correctly.
+                unsigned int numPopStateSetRequired = 0;
+
+                // push the geoset's state on the geostate stack.
+                osg::StateSet* stateset = drawable->getStateSet();
+                if (stateset)
+                {
+                    ++numPopStateSetRequired;
+                    cv->pushStateSet(stateset);
+                }
+
+                osg::CullingSet& cs = cv->getCurrentCullingSet();
+                if (!cs.getStateFrustumList().empty())
+                {
+                    osg::CullingSet::StateFrustumList& sfl = cs.getStateFrustumList();
+                    for(osg::CullingSet::StateFrustumList::iterator itr = sfl.begin();
+                        itr != sfl.end();
+                        ++itr)
+                    {
+                        if (itr->second.contains(bb))
+                        {
+                            ++numPopStateSetRequired;
+                            cv->pushStateSet(itr->first.get());
+                        }
+                    }
+                }
+
+                float depth = FLT_MAX;
+                if(bb.valid())
+                {
+                   for(unsigned int i=0; i<8; ++i)
+                       depth = std::min((float)distance(bb.corner(i), matrix), depth);
```

Comment:
I'm pretty sure I saw a way somewhere in the cull visitor code of finding the nearest BB corner without having to loop through all of them. This isn't the biggest optimisation, but if there's tried and tested OSG code that does the same thing but faster, we might as well do the same thing.

Reply:
could be approx with distance(bb.center,matrix)-BoundingSphere(bb).radius but what's heck...there's not to much transparents

Reply:
...i found the code...

Reply:
..it was for BILLBOARD..:(